sdk/program/src/vote/state/mod.rs

```diff
@@ -492,33 +495,41 @@ impl VoteState {
         }
     }
 
-    /// Deserializes the input buffer into the provided `VoteState`
+    /// Deserializes the input `VoteStateVersions` buffer directly into a provided `VoteState` struct
     ///
-    /// This function exists to deserialize `VoteState` in a BPF context without going above
-    /// the compute limit, and must be kept up to date with `bincode::deserialize`.
+    /// In a BPF context, V0_23_5 is not supported, but in non-BPF, all versions are supported for
+    /// compatibility with `bincode::deserialize`
     pub fn deserialize_into(
         input: &[u8],
         vote_state: &mut VoteState,
     ) -> Result<(), InstructionError> {
-        let minimum_size =
-            serialized_size(vote_state).map_err(|_| InstructionError::InvalidAccountData)?;
-        if (input.len() as u64) < minimum_size {
-            return Err(InstructionError::InvalidAccountData);
-        }
-
         let mut cursor = Cursor::new(input);
 
         let variant = read_u32(&mut cursor)?;
         match variant {
-            // V0_23_5. not supported; these should not exist on mainnet
-            0 => Err(InstructionError::InvalidAccountData),
+            // V0_23_5. not supported for bpf targets; these should not exist on mainnet
+            // supported for non-bpf targets for backwards compatibility
+            0 => {
+                #[cfg(not(target_os = "solana"))]
+                {
+                    *vote_state = bincode::deserialize::<VoteStateVersions>(input)
+                        .map(|versioned| versioned.convert_to_current())
+                        .map_err(|_| InstructionError::InvalidAccountData)?;
+
+                    Ok(())
+                }
+                #[cfg(target_os = "solana")]
+                Err(InstructionError::InvalidAccountData)
+            }
             // V1_14_11. substantially different layout and data from V0_23_5
             1 => deserialize_vote_state_into(&mut cursor, vote_state, false),
             // Current. the only difference from V1_14_11 is the addition of a slot-latency to each vote
             2 => deserialize_vote_state_into(&mut cursor, vote_state, true),
             _ => Err(InstructionError::InvalidAccountData),
         }?;
 
+        // if cursor overruns the input, it produces null bytes and continues to advance `position`
+        // this check ensures we do not accept such a malformed input erroneously
```

Comment:
i added this when i originally wrote the code because i swear i ran into it in testing, but yea youre right. its possible i was still testing with `read` instead of `read_exact`, but even then it just freezes in place (eof or oob if set to oob with `set_position`) so maybe i was mistaken